### ISSUE TITLE
windows: can't start the client app

### ISSUE THREAD

**Issue by selahaddinh** [tailscale/tailscale#747]
**Describe the bug**
I freshly installed the Client for Win 10. It doesn't launch, instead shows an error message. It seems to me like a bug, related to win 10.

**To Reproduce**
Download the client version tailscale-ipn-setup-1.0.4,
I encountered the following error message when I try to run the app,
( also couldn't copy the text inside the alert box, would be cool if there was an option to copy the error log )
```
Tailscale Backend error:
wgengine.NewUserspaceEngine: Error creating interface:
GetStringValueWait(NetCfdInstanceId) failed: Timeout waiting for registry value

log id: (there might be typos...) 
fc2b889dbd894527cf95ba6e458678f0f801cb98a54d6f16c4f0182bd61802e3
```

**Expected behavior**
Wanted tailscale to work in the background, just like it did in Mac and Linux ...

**Screenshots**
![image](https://user-images.githubusercontent.com/6696695/92796198-a4f15480-f3b9-11ea-8be0-8046108cdd35.png)

**Version information:**
Win 10 Education N 1909

**Additional:**
Prolly the problem is related to win 10 edition I am using. It s a free edition distributed only to students afaik and lacks many features.

**Comment by bradfitz** [comment]
> ( also couldn't copy the text inside the alert box, would be cool if there was an option to copy the error log )

In Windows, click the dialog box, then hit Control-C. The text doesn't select so it doesn't look copy-able, but Control-C does work.

**Comment by alexbrainman** [comment]
> Prolly the problem is related to win 10 edition I am using. It s a free edition distributed only to students afaik and lacks many features.

Is there a way for me to download and install that Windows edition to try and reproduce your problem?

Alex

**Comment by selahaddinh** [comment]
I believe the N edition ships only to Europe plus some related regions.

For the Education Edition, many schools have instructional sites on how to download them. i.e. UW: https://itconnect.uw.edu/wares/uware/microsoft/

**Comment by alexbrainman** [comment]
> For the Education Edition, many schools have instructional sites on how to download them. i.e. UW: https://itconnect.uw.edu/wares/uware/microsoft/

@selahaddinh thanks for the link. Unfortunately I cannot download that Windows edition.

![image](https://user-images.githubusercontent.com/9796621/92891494-54323800-f45b-11ea-917d-f93783297e54.png)

Alex

**Comment by alexbrainman** [comment]
@selahaddinh this is the only reasonable suggestion I could find

https://www.mail-archive.com/wireguard@lists.zx2c4.com/msg04397.html

Could you, please check your registry and event logs as requested?

Thank you.

Alex

**Comment by selahaddinh** [comment]
So I reinstalled and updated the windows and also allowed the firewall manually, At least I get to see a slightly different log
`---------------------------
Error
---------------------------
Tailscale service is not running.



safesocket.Connect: dial tcp 127.0.0.1:41112: connectex: No connection could be made because the target machine actively refused it.

---------------------------
Tamam   
---------------------------
`

**Comment by alexbrainman** [comment]
> safesocket.Connect: dial tcp 127.0.0.1:41112: connectex: No connection could be made because the target machine actively refused it.

Tailscale for Windows consists of Windows service and system tray application. These 2 talk to each other. This message indicates that your system tray app cannot talk to the Windows service part. I suspect your Tailscale Windows service crashed or not started for some reason.

Perhaps try to install latest Tailscale version

https://pkgs.tailscale.com/unstable/

Alex

**Comment by selahaddinh** [comment]
I start to think that the issue is related to: https://github.com/tailscale/tailscale/issues/524 
Because the error is somewhat similar but I am not sure tho. It works well on my other pc as well.

**Comment by alexbrainman** [comment]
@selahaddinh 

Did you actually tried to install latest version for Windows from

https://pkgs.tailscale.com/unstable/

? What happened? What did you see? Please, reboot your computer after installation is finished.

> I start to think that the issue is related to: #524
> Because the error is somewhat similar but I am not sure tho. It works well on my other pc as well.

Which error on  #524 is similar to yours?

Thank you.

Alex

**Comment by selahaddinh** [comment]
### Did you actually tried to install latest version for Windows from
### What happened? What did you see? Please, reboot your computer after installation is finished.

I am sorry for not sharing the details,
I indeed installed the unstable version and the same problem persisted :/
More specifically, I uninstalled, restarted, installed this version https://pkgs.tailscale.com/unstable/tailscale-ipn-setup-1.1.335.exe then restarted. 
![image](https://user-images.githubusercontent.com/6696695/93027009-dfa1f980-f612-11ea-8046-6a02fac19f28.png)
I ignore the error in the left dialog box since the service is not working and the corresponding port is not responding,
 So I am here with this error: 
```
---------------------------
Tailscale backend error:

wgengine.NewUserspaceEngine: Error creating interface: GetStringValueWait(NetCfgInstanceId) failed: Timeout waiting for registry value

logid: fc2b889dbd894527cf95ba6e458678f0f801cb98a54d6f16c4f0182bd61802e3
---------------------------
```
### Which error on #524 is similar to yours?
I mean the logIDs are different, but the error strings are similar. But I have no further supports for the claim 👎  
https://github.com/tailscale/tailscale/issues/524#issuecomment-665010233
https://github.com/tailscale/tailscale/issues/524#issuecomment-665455465

At this point I am just tired of windows, so I just connect to another local Linux machine as Tailscale endpoint. GL on dealing with WinApi.

**Comment by selahaddinh** [comment]
Okey just a small side note, I forked wireguard-go to log the error a little bit more in detail,

https://github.com/selahaddinh/wireguard-go/commit/110efef1b29cdabcf8805fb9ea027310eb1b0577

here's the output from Admin Powershell, 
`go run .\tailscaled.go --state 0` I hope that, this is the right way to call to tailscaled for windows powershell,

![image](https://user-images.githubusercontent.com/6696695/93031190-98773100-f631-11ea-8e83-19913475d3c4.png)

**Comment by alexbrainman** [comment]
> I am sorry for not sharing the details,

No worries.

> More specifically, I uninstalled, restarted, installed this version https://pkgs.tailscale.com/unstable/tailscale-ipn-setup-1.1.335.exe

Good. Just making sure you don't use old version and hitting bugs that are already fixed.

> So I am here with this error:
> 
> ```
> ---------------------------
> Tailscale backend error:
> 
> wgengine.NewUserspaceEngine: Error creating interface: GetStringValueWait(NetCfgInstanceId) failed: Timeout waiting for registry value
> 
> logid: fc2b889dbd894527cf95ba6e458678f0f801cb98a54d6f16c4f0182bd61802e3
> ```

I see what you mean. Thank you for explaining. Did you read

https://www.mail-archive.com/wireguard@lists.zx2c4.com/msg04397.html

? Can you check your registry to make sure this key

`HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Class\{4d36e972-e325-11ce-bfc1-08002be10318}\`

is present. This is how my registry looks like

![image](https://user-images.githubusercontent.com/9796621/93067852-10d80380-f6bf-11ea-8d96-86f227bf07fa.png)

In particular I have a bunch of subkeys named like `000` and `001` and similar, and each subkey has value named `NetCfgInstanceId`. That value is read by the program line that fails for you. Do you have these keys and values?

Also the

https://www.mail-archive.com/wireguard@lists.zx2c4.com/msg04397.html

mentions that there might be errors in `Event Viewer`. I opened this key

`%SystemRoot%\System32\Winevt\Logs\Microsoft-Windows-DeviceSetupManager%4Admin.evtx`

but I don't have any errors.

![image](https://user-images.githubusercontent.com/9796621/93068430-c73be880-f6bf-11ea-9899-417c0049fe1c.png)

What about you?

> At this point, I am just tired of windows, it's better to connect thru another local Linux machine as Tailscale endpoint.

Fair enough. If you don't want to debug this, I understand. Feel free to ignore my questions.

> I have no experience with WinApi, posting these in case it is meaningful to some

Unfortunately these are not helpful. This bit of code waits for the registry key / value I described above to be present. And the code waits for 10 seconds. And then it gives up. But we already know all that. We need to know why the key is not appearing? Unfortunately I have no clue.

Alex

**Comment by selahaddinh** [comment]
Thank you, I'll check the regedit and `Event Viewer` tomorrow hopefully. if it's not bothering you, let's keep debugging then, maybe this'll be a reference issue for future.

**Comment by selahaddinh** [comment]
`Can you check your registry to make sure this key ... {4d36e972-e325-11ce-bfc1-08002be10318}\...`
Yes, they exist.
![image](https://user-images.githubusercontent.com/6696695/93254496-19086f80-f7a1-11ea-8564-98e67c3d04fc.png)
Particularly there are 2 records named as wintun and related with the wireguard:
![image](https://user-images.githubusercontent.com/6696695/93255502-b31ce780-f7a2-11ea-8f9f-599045f76ca1.png)
![image](https://user-images.githubusercontent.com/6696695/93255758-02631800-f7a3-11ea-8a5b-d8227e13cd0e.png)
`... mentions that there might be errors in Event Viewer`
I have checked but mine was no different, no visible error log in the near past. (`Bilgi` = `Information`)
![image](https://user-images.githubusercontent.com/6696695/93255197-3853cc80-f7a2-11ea-82a4-d38f9956a42c.png)

**Comment by bradfitz** [comment]
I see your computer's in Turkish. Perhaps we have some bug where we're assuming English localization values somewhere?

**Comment by selahaddinh** [comment]
it is running well on my Turkish localized laptop.

**Comment by alexbrainman** [comment]
> But I compared the wintun device records between the two machines and found out not working one lacks

The code that fails is looking for a different value - `NetCfgInstanceId`. Unfortunately I do not know the key path. Windows API that is used in this code returns already open registry key handle, so we do not know the key path. There is no "official" Windows API to convert key handle into key path, but maybe we could find something anyway. I am not sure how that might help us to know the key path.

As to the fact that you don't have `ComponentId` value in one of your keys - it might be related to your bug. Perhaps it is used to find key with `NetCfgInstanceId` that we are missing. I do not know.

I am not sure why you do not have `ComponentId`. I was thinking that whole key tree related to our driver is installed by Tailscale installer. So I deleted whole tree rooted at

`[HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Class\{4d36e972-e325-11ce-bfc1-08002be10318}]`

But Tailscale installer did not restore the tree. And my network stopped working completely. Even after I added the key manually (by merging registry backup text file I created before deleting), my network did not work.

I had to "Reset network" whatever that mean and reboot after that, to make my network work again. Don't do that to yourself.

I am out of ideas for the moment.

Alex

**Comment by selahaddinh** [comment]
After this, I investigated with OpenVPN with wintun adapter too, in both Tailscale and OpenVpn the Wintun Device was giving errors( OpenVpn didn't even finished the installation and instead just gave error). At this point, I think the issue is out of the scope. So I am closing the issue, Thank you all for your help.

**Comment by alexbrainman** [comment]
> After this, I investigated with OpenVPN with wintun adapter too, in both Tailscale and OpenVpn the Wintun Device was giving errors( OpenVpn didn't even finished the installation and instead just gave error). At this point, I think the issue is out of the scope. 

I would be still interested to know what is wrong with the driver. Not sure how to go about it. Maybe you can show the error you saw from OpenVpn installer. If you don't mind.

> So I am closing the issue, Thank you all for your help.

No worries. Sorry I could not help you get Tailscale running.

Alex

**Comment by selahaddinh** [comment]
it's hers (tailscale's) loss, anyway...
`Maybe you can show the error you saw from OpenVpn installer`
Sure, I'll try to show the logs for the OpenVpn and the state of Wintun for both cases when I get to PC.

**Comment by selahaddinh** [comment]
![image](https://user-images.githubusercontent.com/6696695/93705299-ccb48b00-fb24-11ea-862b-3d8479e2be31.png)
![image](https://user-images.githubusercontent.com/6696695/93705312-e5bd3c00-fb24-11ea-9dad-7b529325a700.png)
It is not logging what causes the error at the client.

For the drivers case:
This time I installed a newer version of unstable release (it warns for Trojan btw )
In device manager, there is a warning sign for Tailscale Tunnel, it disappears after the service error dialog box opens.
![image](https://user-images.githubusercontent.com/6696695/93705536-f373c100-fb26-11ea-8114-904a54e09e42.png)
![image](https://user-images.githubusercontent.com/6696695/93705525-e0f98780-fb26-11ea-83af-3e90f0d27d36.png)

**Comment by alexbrainman** [comment]
@selahaddinh thanks for posting your error message. It is not much, but hopefully it will help future users.

Alex

**Comment by divi2323** [comment]
I was getting this pretty consistently with both Windows 2016 Server as well as Many versions of Windows 10.  One thing I notice in the screenshots above, some of them contain references to hyper-v.  I removed hyper-v from my systems with this issue and it was able to install properly.  I would assume that when it goes to create the virtual network adapter, it's not able to check for something, and the default hyper-v switch virtual interface in the network adapters page caused it to fail.

Removing hyper-v fixed it for me.  Readding hyper-v into the system would break it again, but only during install.  To fix it (at least for what I experienced), remove hyper-v, or just the virtual switch adapter it creates, then install tailscale, configure, and then reenable, or reinstall hyper-v.

I would assume this would creep up on other systems where a virtual interface is created by an application, and blocks tailscale's ability to do so.  For example openVPN was used above, as well as some other VPN software.  It likely creates a networking device like this and may either fix it, or allow tailscale to work around it.

Hope this helps someone.
